### PR TITLE
Skill perfect item template

### LINKED ISSUE #1196: DB restructuring - world - skill_perfect_item_template
This is part of: https://github.com/azerothcore/azerothcore-wotlk/issues/1118

**General note**: not all structure differences can or should be imported. Every change must be first verified. If a structure change is not suitable, we just close the issue.

---

https://github.com/TrinityCore/TrinityCore/commit/4771740d424bea45a12e9d47d7919f2f44929226

```diff
+-- Dumping structure for table tc_world.skill_perfect_item_template
+DROP TABLE IF EXISTS `skill_perfect_item_template`;
+CREATE TABLE IF NOT EXISTS `skill_perfect_item_template` (
+  `spellId` mediumint(8) unsigned NOT NULL DEFAULT '0' COMMENT 'SpellId of the item creation spell',
+  `requiredSpecialization` mediumint(8) unsigned NOT NULL DEFAULT '0' COMMENT 'Specialization spell id',
+  `perfectCreateChance` float NOT NULL DEFAULT '0' COMMENT 'chance to create the perfect item instead',
+  `perfectItemType` mediumint(8) unsigned NOT NULL DEFAULT '0' COMMENT 'perfect item type to create instead',
+  PRIMARY KEY (`spellId`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8 ROW_FORMAT=FIXED COMMENT='Crafting Perfection System';
```

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/68093669-db-restructuring-world-skill_perfect_item_template?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Ercules76**
##### CHANGES PROPOSED:

-  Create new table for perfect item craft
-  Implement inside core the function to manage proc

###### ISSUES ADDRESSED:

Closes #1196 #1195 

##### TESTS PERFORMED:

The build has been successfully completed on Windows 10, compiled at 64bit

##### HOW TO TEST THE CHANGES:

Test with a PG and jewel skill, you must have the spell Gem Perfection (id: 55534), if look at the new table a specific spell id is possible know the proc percentage and the perfect item that would be won, 

ex:
spell (53831 [Bold Bloodstone]) if proc win (41432 [Perfect Bold Bloodstone])

##### KNOWN ISSUES AND TODO LIST:

I suggest this test:
- choose a spellId in table skill_perfect_item_template and change the chance to 100%, if you are in game is possible reload the now table with the command:

.reload skill_extra_item_template

the command reload both table skill_extra_item_template and skill_perfect_item_template (the second  latch onto skill_extra_item_template to avoid changing command IDs)
- with the proc change at 100% unlearn the spell 55534 and test that no perfect items are crafted.

- restore the right value in DB

##### Target branch(es):

Master

Special thanks to Treeston <3

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
codestyle looks good, well done @Ercules76 

let's wait for testers to check this

**Comment by Helias** [comment]
While adding the SQL file I get:
```
ERROR 1314 (0A000) at line 3743: LOCK is not allowed in stored procedures
```
OS: Ubuntu 18.04
mysql: 5.7.25

To get a similar results instead of **LOCK** and **UNLOCK** tables you can use **START TRANSACTION;** and **COMMIT;**
reference: https://stackoverflow.com/questions/51137633/1314-lock-is-not-allowed-in-stored-procedures

Change this for cross-compatibility.


About the in-game effect, it works!
I tested it with Bold Bloodstone and Bright Bloodstone.
immagine

(I didn't cheat modifying 100% percentage).

So change the LOCK/UNLOCK tables and the PR could be ready to merge.

File: data/sql/updates/pending_db_world/rev_1552232377246845400.sql [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+CREATE TABLE `skill_perfect_item_template` (
+	`spellId` mediumint(8) unsigned NOT NULL DEFAULT '0' COMMENT 'SpellId of the item creation spell',
+	`requiredSpecialization` mediumint(8) unsigned NOT NULL DEFAULT '0' COMMENT 'Specialization spell id',
+	`perfectCreateChance` float NOT NULL DEFAULT '0' COMMENT 'chance to create the perfect item instead',
+	`perfectItemType` mediumint(8) unsigned NOT NULL DEFAULT '0' COMMENT 'perfect item type to create instead',
+	PRIMARY KEY (`spellId`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8 ROW_FORMAT=FIXED COMMENT='Crafting Perfection System';
+
+LOCK TABLES `skill_perfect_item_template` WRITE;
```
Comment (Viste) [inline data/sql/updates/pending_db_world/rev_1552232377246845400.sql]: Please change LOCK to START TRANSACTION

File: data/sql/updates/pending_db_world/rev_1552232377246845400.sql [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+(53867,55534,20,41454), -- Infused
+(53865,55534,20,41455), -- Mysterious
+(53870,55534,20,41456), -- Puissant
+(53863,55534,20,41457), -- Purified
+(53868,55534,20,41458), -- Regal
+(53864,55534,20,41459), -- Royal
+(53860,55534,20,41460), -- Shifting
+(53859,55534,20,41461), -- Sovereign
+(53861,55534,20,41462); -- Tenuous
+UNLOCK TABLES;
```
Comment (Viste) [inline data/sql/updates/pending_db_world/rev_1552232377246845400.sql]: Please change UNLOCK to COMMIT

Comment (Ercules76) [reply data/sql/updates/pending_db_world/rev_1552232377246845400.sql]: edited

File: data/sql/updates/pending_db_world/rev_1552232377246845400.sql (hunk 1 above)
Comment (Ercules76) [reply data/sql/updates/pending_db_world/rev_1552232377246845400.sql]: edited

**Pull request merged**